app/services/state_token_encoder_service.rb

```diff
@@ -0,0 +1,66 @@
+# frozen_string_literal: true
+
+require 'json_web_token'
+
+##
+# Service which encodes current sign-in state into a JWT token
+class StateTokenEncoderService < ApplicationService
+  Response = Data.define(:status, :body)
```

Comment:
Ooooh shiny new `Ruby` classes

Reply:
I'm gonna have to research usage of this a bit, I think this accomplishes what I use a lot of `OpenStruct` stuff for (and various tiny, tiny classes I create because I want some enforced form to the data and don't just want to pass around formless hashes)

Reply:
Yeah, I started playing around with it recently and liked it so much I refactored all of my service objects in a side-project to return `Data` objects. I like that the response is immutable yet the individual values within the response are still mutable, allowing for a high degree of confidence in the integrity of the response while also allowing for further processing of the data returned in the response.

Reply:
`Response` itself is a class here, right? I see that it gets called as a child of `StateTokenEncorderService` in the OAuth controller spec; is it the case that you can just declare a class as a variable like that and have it be callable from within and externally?

Reply:
The `Response` class is an example of a value object which was introduced in Ruby 3.2 as the `Data` class. I like to think of them as an immutable-ish `Struct`. You can read more about them here: https://docs.ruby-lang.org/en/3.2/Data.html

Reply:
As for the encapsulation piece, keeping the `Response` public represents no more risk than defining an error class within another class. It also makes it easy to use doubles to mock the results of calling the class without all the additional test set-up, allowing for cleaner boundaries in your specs.